File: 28.implement-str-str.cpp
/*
 * @lc app=leetcode id=28 lang=cpp
 *
 * [28] Implement strStr()
 */

#include <cassert>
#include <cstring>
#include <string>

using namespace std;

// @lc code=start
class Solution {
 public:
  int strStr(string haystack, string needle) {
    size_t hlen = haystack.length();
    size_t nlen = needle.length();

    if (nlen == 0) {
      return 0;
    }

    if (hlen < nlen) {
      return -1;
    }

    for (u_int i = 0; i + nlen <= hlen; i++) {
      for (u_int j = 0; j < nlen; j++) {
        for (u_int k = 0; j + k < nlen; k++) {
          if (haystack[i + k] != needle[j + k]) {
            goto NEXT_LOOP;
          }
        }
        return i;
      }
    NEXT_LOOP:;
    }
    return -1;
  }
};
// @lc code=end

int main() {
  Solution s;
  assert(s.strStr("a", "a") == 0);
  assert(s.strStr("aaaaa", "bba") == -1);
  assert(s.strStr("hello", "ll") == 2);
  assert(s.strStr("", "bba") == -1);
  assert(s.strStr("aaaaa", "") == 0);
  assert(s.strStr("aaaaa", "a") == 0);
  assert(s.strStr("aaaaa", "aaaaaa") == -1);
  printf("OK\n");
}

File: 28.implement-str-str.rs
/*
 * @lc app=leetcode id=28 lang=rust
 *
 * [28] Implement strStr()
 */

struct Solution;

// @lc code=start
impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        let hlen = haystack.len();
        let nlen = needle.len();
        if needle.is_empty() {
            return 0;
        }
        if hlen < nlen {
            return -1;
        }
        for i in 0..(hlen - nlen + 1) {
            if haystack[i..i + nlen] == needle {
                return i as i32;
            }
        }
        -1
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::str_str("hello".to_string(), "ll".to_string()), 2);
}

#[test]
fn test2() {
    assert_eq!(
        Solution::str_str("aaaaa".to_string(), "bba".to_string()),
        -1
    );
}

#[test]
fn test3() {
    assert_eq!(Solution::str_str("a".to_string(), "a".to_string()), 0);
    assert_eq!(Solution::str_str("".to_string(), "".to_string()), 0);
}
